elyra/cli/pipeline_app.py

```diff
@@ -37,13 +42,29 @@ def _get_runtime_type(runtime_config: str) -> str:
         raise click.ClickException(f'Invalid runtime configuration: {runtime_config}\n {e}')
 
 
-def _validate_pipeline_file(pipeline_file):
+def _validate_pipeline_runtime(primary_pipeline, runtime):
+    """
+    Generic pipelines do not have a persisted runtime type, and can be run on any runtime
+    Runtime specific pipeline have a runtime type, and con only be run on matching runtime
+    """
+    is_valid = False
+    pipeline_runtime = primary_pipeline["app_data"].get("runtime")
+    if not pipeline_runtime:
+        is_valid = True
+    else:
+        if runtime and pipeline_runtime == runtime:
+            is_valid = True
+
+    return is_valid
+
+
+def _validate_pipeline_file_extension(pipeline_file: str):
     extension = os.path.splitext(pipeline_file)[1]
     if extension != '.pipeline':
         raise click.ClickException('Pipeline file should be a [.pipeline] file.\n')
 
 
-def _preprocess_pipeline(pipeline_path, runtime, runtime_config):
+def _preprocess_pipeline(pipeline_path: str, runtime: str, runtime_config: str):
     pipeline_path = os.path.expanduser(pipeline_path)
```

Comment:
Is this necessary?  This will essentially make `pipeline_path` absolute, yet it's getting joined to another directory in the next line.

Reply:
this expand paths expressed as "~/......"

Reply:
Correct, which results in something like `/home/alice/...`.  Then, (_I thought_) the next line `os.path.join(os.getcwd(), pipeline_path)` would result in: `/home/alice/home/alice/...` (for example), but I didn't realize `join` does this: _If a component is an absolute path, all previous components are thrown away and joining continues from the absolute path component._ so the current working directory will not be prefixed.

So it seems like things will "fall out" correctly, although I'm not sure that was the intention when this was written.